src/macros-by-example.md

```diff
@@ -24,149 +24,404 @@
-> &nbsp;&nbsp; | `$` `(` _MacroMatch_<sup>+</sup> `)` _MacroRepSep_<sup>?</sup> _MacroKleeneOp_
+> &nbsp;&nbsp; | `$` `(` _MacroMatch_<sup>+</sup> `)` _MacroRepSep_<sup>?</sup> _MacroRepOp_
 >
 > _MacroFragSpec_ :\
 > &nbsp;&nbsp; &nbsp;&nbsp; `block` | `expr` | `ident` | `item` | `lifetime`\
 > &nbsp;&nbsp; | `meta` | `pat` | `path` | `stmt` | `tt` | `ty` | `vis`
 >
 > _MacroRepSep_ :\
-> &nbsp;&nbsp; [_Token_]<sub>_except delimiters and kleene operators_</sub>
+> &nbsp;&nbsp; [_Token_]<sub>_except delimiters and repetition operators_</sub>
 >
-> _MacroKleeneOp_<sub>2015</sub> :\
-> &nbsp;&nbsp; `*` | `+`
->
-> _MacroKleeneOp_<sub>2018+</sub> :\
-> &nbsp;&nbsp; `*` | `+` | `?`
+> _MacroRepOp_<sub>2018+</sub> :\
+> &nbsp;&nbsp; `*` | `+` | `?`<sub>2018+</sub>
 >
 > _MacroTranscriber_ :\
 > &nbsp;&nbsp; [_DelimTokenTree_]
 
 `macro_rules` allows users to define syntax extension in a declarative way.  We
 call such extensions "macros by example" or simply "macros".
 
-Macros can expand to expressions, statements, items, types, or patterns.
-
-The macro expander looks up macro invocations by name, and tries each macro
-rule in turn. It transcribes the first successful match. Matching and
-transcription are closely related to each other, and we will describe them
-together.
-
-The macro expander matches and transcribes every token that does not begin with
-a `$` literally, including delimiters. For parsing reasons, delimiters must be
-balanced, but they are otherwise not special.
-
-In the matcher, `$` _name_ `:` _designator_ matches the nonterminal in the Rust
-syntax named by _designator_. Valid designators are:
-
-* `item`: an [_Item_]
-* `block`: a [_BlockExpression_]
-* `stmt`: a [_Statement_] without the trailing semicolon
-* `pat`: a [_Pattern_]
-* `expr`: an [_Expression_]
-* `ty`: a [_Type_]
-* `ident`: an [IDENTIFIER_OR_KEYWORD]
-* `path`: a [_TypePath_] style path
-* `tt`: a [_TokenTree_]&nbsp;(a single [token] or tokens in matching delimiters `()`, `[]`, or `{}`)
-* `meta`: a [_MetaItem_], the contents of an attribute
-* `lifetime`: a [LIFETIME_TOKEN]
-* `vis`: a [_Visibility_] qualifier
-* `literal`: matches `-`<sup>?</sup>[_LiteralExpression_]
+Each macro by example has a name, and one or more _rules_. Each ruleh as two
+parts: a _matcher_, describing the syntax that it matches, and a _transcriber_,
+describing the syntax that will replace a successfully matched invocation. Both
+the matcher and the transcriber must be surrounded by delimiters. Macros can
+expand to expressions, statements, items, types, or patterns.
+
+When a macro is invoked, the macro expander looks up macro invocations by name,
+and tries each macro rule in turn. It transcribes the first successful match; if
+this results in an error, then future matches are not tried. When matching, no
+lookahead is performed; if the compiler cannot umambiguously determine how to
+parse the macro invocation one token at a time, then it is an error. In the
```

Comment:
I'm not sure whether "one token at time" includes https://github.com/rust-lang/rust/pull/42913 or not.